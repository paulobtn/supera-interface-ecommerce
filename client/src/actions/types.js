// Adiciona item ao carrinho de compras
export const ADD_TO_CART = "add_to_cart";

