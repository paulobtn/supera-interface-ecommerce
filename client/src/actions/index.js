import { ADD_TO_CART } from "./types";

/* Adiciona produto ao carrinho de compras e sua quantidade.
 * Caso jÃ¡ exista, apenas soma a quantidade
 */
export const addToCart = (qtd, data) => {
  return{
    type: ADD_TO_CART,
    payload: {qtd, data}
  }
};
