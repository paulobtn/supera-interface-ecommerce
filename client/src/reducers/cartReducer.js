import { ADD_TO_CART } from '../actions/types';

const freeShippingValue = 250;
const shippingValue = 10;

//exportando para fins de teste
export const initState = { 
  shoppingCart:[], //[{quantity, item{id,name,price,score,image}}]
  quantity: 0,
  subtotal: 0,
  shipping: 0, //apenas adicionado ao total se freeShipping for falso
  freeShipping: false,
  total: 0
}

/* stateAddToCart retorna o novo estado ao adicionar o produto ao
 * carrinho */
const stateAddToCart = (state, action) => {

  /* dado e quantidade do payload para não termos que escrever
   * action.payload varias vezes */
  let {data, quantity} = action.payload;

  /* faz uma deep copy do state, note que apenas 
   * newState = {...state}
   * iria copiar a referência de shoppingCart e não queremos
   * alterar state, apenas retornar um novo
   */
  let newState = JSON.parse(JSON.stringify(state));

  // verifica se o produto já existe
  let product = newState.shoppingCart.find( (product) => {
    return product.data.id === data.id
  });
  
  // Adiciona produto ao carrinho ou incrementa quantidade
  if(product){
    newState
      .shoppingCart[newState.shoppingCart.indexOf(product)]
      .quantity += quantity;
  } else{
    newState.shoppingCart = newState.shoppingCart.concat(action.payload);
  }
    
  // atualiza quantidade e subtotal
  newState.quantity += quantity;
  newState.subtotal += data.price * quantity;

  // atualiza fretes
  newState.shipping += quantity*shippingValue;
  if(newState.subtotal > 250){
    newState.freeShipping = true;
  }

  // atualiza total
  newState.total = newState.subtotal;
  if(!newState.freeShipping){
    newState.total += newState.shipping;
  }
  return newState;
}

const cartReducer = (state = initState, action) => {

    switch(action.type){
        case ADD_TO_CART:
          return stateAddToCart(state, action);
        default:
            return state;
    }
}

export default cartReducer;
