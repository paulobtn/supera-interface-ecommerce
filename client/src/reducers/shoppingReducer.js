import {FETCH_GAMES, SET_ERROR} from '../actions/types';

const initState = {
    items: [],
    shoppingCart:[],
    error: null
}

const shoppingReducer = (state = initState, action) => {

    switch(action.type){
        case FETCH_GAMES:
        /* Adiciona todos os jogos no campo items do 
         * redux store */
            return {
              ...state,
              error: null,
              items: action.payload
            }
        case SET_ERROR:
          return {
            ...state,
            error: action.payload
          }
        default:
            return state;
    }
}

export default shoppingReducer;
