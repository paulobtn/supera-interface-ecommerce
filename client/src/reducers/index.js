import { combineReducers } from "redux";
import shoppingReducer from "./shoppingReducer";

export default combineReducers({
  shopping: shoppingReducer
});
