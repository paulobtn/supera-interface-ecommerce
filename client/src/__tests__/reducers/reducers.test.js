import configureMockStore from 'redux-mock-store';

import cartReducer from '../../reducers/cartReducer';
import {initState} from '../../reducers/cartReducer';
import { ADD_TO_CART, REMOVE_FROM_CART } from "../../actions/types";


describe('Testa reducer', () => {
  
  describe('cartReducer', () => {

    it('retornando o estado inicial', () => {
      expect(cartReducer(undefined, {})).toMatchSnapshot();
    });

    it('retornando estado correto após adicionar um jogo', () => {

      let game = {
        "id": 201,
        "name": "Call Of Duty Infinite Warfare",
        "price": 49.99,
        "score": 80,
        "image": "call-of-duty-infinite-warfare.png"
      };
      
      expect (cartReducer(initState, {
        type: ADD_TO_CART,
        payload: {
          quantity: 10,
          data: game
        }
      })).toMatchSnapshot();  

    })

    it('retornando estado correto após remover um jogo', () => {

      let game = {
        "id": 201,
        "name": "Call Of Duty Infinite Warfare",
        "price": 49.99,
        "score": 80,
        "image": "call-of-duty-infinite-warfare.png"
      };
      
      // adiciona no carrinho
      let state = cartReducer(initState, {
        type: ADD_TO_CART,
        payload: {
          quantity: 1,
          data: game
        }
      });  

      // testa a remoção
      expect (cartReducer(initState, {
        type: REMOVE_FROM_CART,
        payload: {
          id: 201
        }
      })).toMatchSnapshot();  

    })
  });
});
