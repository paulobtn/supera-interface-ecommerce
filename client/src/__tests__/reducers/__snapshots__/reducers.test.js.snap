// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testa reducer cartReducer retornando estado correto após adicionar um jogo 1`] = `
Object {
  "freeShipping": true,
  "quantity": 10,
  "shipping": 100,
  "shoppingCart": Array [
    Object {
      "data": Object {
        "id": 201,
        "image": "call-of-duty-infinite-warfare.png",
        "name": "Call Of Duty Infinite Warfare",
        "price": 49.99,
        "score": 80,
      },
      "quantity": 10,
    },
  ],
  "subtotal": 499.90000000000003,
  "total": 499.90000000000003,
}
`;

exports[`Testa reducer cartReducer retornando estado correto após remover um jogo 1`] = `
Object {
  "freeShipping": false,
  "quantity": 0,
  "shipping": 0,
  "shoppingCart": Array [],
  "subtotal": 0,
  "total": 0,
}
`;

exports[`Testa reducer cartReducer retornando o estado inicial 1`] = `
Object {
  "freeShipping": false,
  "quantity": 0,
  "shipping": 0,
  "shoppingCart": Array [],
  "subtotal": 0,
  "total": 0,
}
`;
